contributors/design-proposals/key-set-as-merge-key.md

```diff
@@ -0,0 +1,417 @@
+
+## Background
+
+Strategic Merge Patch is covered in this [doc](https://github.com/kubernetes/community/blob/master/contributors/devel/strategic-merge-patch.md).
+In Strategic Merge Patch, Merge Key is the key to identify the entries in the list of non-primitive types.
+It must always be present and unique to perform the merge on the list of non-primitive types,
+and will be preserved.
+
+The merge key exists in the struct tag (e.g. in [types.go](https://github.com/kubernetes/kubernetes/blob/5a9759b0b41d5e9bbd90d5a8f3a4e0a6c0b23b47/pkg/api/v1/types.go#L2831))
+and the [OpenAPI spec](https://github.com/kubernetes/kubernetes/blob/master/api/openapi-spec/swagger.json).
+
+## Motivation
+
+The current implementation requires a single field that uniquely identifies each element in a list.
+For some element Kinds, the identity is defined using multiple fields.
+An [example](https://github.com/kubernetes/kubernetes/issues/39188) is the service.spec.ports,
+which is identified by both `protocol` and `port`.
+
+As a result we need to also support a set of keys as a Merge Key.
+This will not only fix the APIs that cannot be effectively identified by one single field
+but also benefit new APIs when we want to use multi-field merge key for them.
+
+## Proposed Change
+
+### API Change
+
+We introduce a separate set of merge keys, called the recommended merge keys,
+for all of the fields using merge key.
+
+It will be in a new struct tag with key `recommendedPatchMergeKey`.
+The value will be merge keys seperated by ",", i.e. `recommendedPatchMergeKey:"<key1>,<key2>,<key3>"`.
+
+We will keep the old merge key to keep backward compatibility, and we call them the default merge key.
+
+- For API resources that cannot be effectively merged with a single merge key,
+we will introduce additional merge keys in the recommended merge keys.
+- For the others, the recommended merge keys will be the same as the default merge key.
+
+Requirements for the recommended merge keys to keep backward compatibility:
+- the recommended merge keys must have the default merge key
+- the default merge key must be the first one in the recommended merge keys, i.e. `recommendedPatchMergeKey:"<default-merge-key>,<additional-key1>,<additional-key2>"`
+- the default merge key must be present in the patch, i.e. the first key in the recommended merge keys must be used.
+
+E.g. [`Ports` in `ServiceSpec`](https://github.com/kubernetes/kubernetes/blob/c51efa9ba0929a643544078d5c182ba75e4b4087/pkg/api/v1/types.go#L2825-L2831).
+```go
+type ServiceSpec struct {
+  // add recommendedPatchMergeKey "port,protocol"
+  Ports []ServicePort `patchMergeKey:"port" recommendedPatchMergeKey:"port,protocol" ...`
+  ...
+}
+```
+
+We also need to introduce a new optional directive `$patchMergeKey` in the patch.
+This directive contains a list of strings which are the merge keys used in this patch.
+
+An example patch will look like:
+```yaml
+list:
+- $patchMergeKey:
```

Comment:
in this example, `$patchMergeKey` is included in the first item's definition... is that intended?